torch_xla/csrc/ir.h

```diff
@@ -0,0 +1,144 @@
+
+  std::string ToString() const;
+
+  // The node using the output of the node this use belongs to.
+  Node* node = nullptr;
+  // The operand index, within node's operands, which this use refers to.
+  size_t operand_index = 0;
+  // The index within output the user node refers to.
+  size_t index = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& stream, const Use& use) {
+  stream << use.ToString();
+  return stream;
+}
+
+// Represents a specific value produced by a node. Since the output of a node
+// can be composed by multiple values, the node+index coordinates fully qualify
+// each single output.
+struct Value {
+  struct Hasher {
+    size_t operator()(const Value& value) const {
+      size_t h = reinterpret_cast<std::ptrdiff_t>(value.node);
+      return h ^ (h >> 11) ^ value.index;
+    }
+  };
+
+  Value() = default;
+  explicit Value(Node* node, size_t index = 0) : node(node), index(index) {}
+
+  bool operator==(const Value& rhs) const {
+    return node == rhs.node && index == rhs.index;
+  }
+  bool operator!=(const Value& rhs) const {
+    return !operator==(rhs);
+  }
+
+  std::string ToString() const;
+
+  // The node providing the value.
+  Node* node = nullptr;
+  // The index in the node's output this value refers to.
+  size_t index = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& stream, const Value& value) {
+  stream << value.ToString();
+  return stream;
+}
+
+using ValueSet = std::unordered_set<Value, Value::Hasher>;
+
+template <typename T>
+using ValueMap = std::unordered_map<Value, T, Value::Hasher>;
+
+// A node in the graph. Nodes for operations which requires extra data to be
+// stored for lowering, should inherit from this class and add operation
+// specific member there. For example, a constant might create a new
+// NodeConstant class (inheriting from Node) with an extra xla::Literal field,
+// or a tensor value might create a new NodeTensor with computation client data
+// handle in it.
+class Node {
+ public:
+  // Creates a new node with the given op name. The op name is a unique
+  // identifier for the operation, which in the PyTorch case will be the full
+  // operation signature which is currently used throughout the code base.
+  // The num_outputs tells how many outputs a given operation generates.
+  explicit Node(std::string op, size_t num_outputs = 1);
+
+  virtual ~Node();
+
+  const std::string& op() const { return op_; }
+
+  size_t num_outputs() const { return num_outputs_; }
+
+  const std::vector<Value>& operands() const { return operands_as_values_; }
+
+  const std::set<Use>& uses() const { return uses_; }
+
+  // Adds node's index output number as operand.
+  void AddOperand(NodePtr node, size_t index = 0);
+
+  void ReplaceOperand(size_t operand_no, NodePtr node, size_t index = 0);
+
+  void ReplaceAllUsesWith(NodePtr node, size_t index = 0);
+
+  virtual std::string ToString() const;
+
+  virtual xla::XlaOp Lower(LoweringContext* loctx) const;
```

Comment:
This will likely need to be a list of `xla::XlaOp` instead. Think of backward convolution, it returns three gradients (for input, kernel and bias).